src/persist-txn/src/txn_read.rs

```diff
@@ -321,19 +340,85 @@ enum WaitTs<T> {
 }
 
 #[derive(Debug)]
-struct TxnsReadTask<T: Timestamp + Lattice + Codec64, C: TxnsCodec = TxnsCodecDefault> {
+struct TxnsReadTask<T: Timestamp + Lattice + Codec64> {
     rx: mpsc::Receiver<TxnsReadCmd<T>>,
-    cache: TxnsCache<T, C>,
+    cache: TxnsCacheState<T>,
+    pending_waits: Vec<PendingWait<T>>,
+    frontier: T,
+}
+
+/// A pending "wait" notification that we will complete once the frontier
+/// advances far enough.
+#[derive(Debug)]
+struct PendingWait<T: Timestamp + Lattice + Codec64> {
+    ts: WaitTs<T>,
+    tx: Option<oneshot::Sender<()>>,
 }
 
-impl<T, C> TxnsReadTask<T, C>
+impl<T: Timestamp + Lattice + Codec64> PendingWait<T> {
+    /// Returns `true` if this [PendingWait] is completed.
+    ///
+    /// A pending wait is completed when the frontier advances far enough or
+    /// when the receiver side hangs up.
+    fn maybe_complete(&mut self, frontier: &T) -> bool {
+        if self.tx.is_none() {
+            // Already completed!
+            return true;
+        }
+
+        match &self.ts {
+            WaitTs::GreaterEqual(ts) => {
+                if frontier >= ts {
+                    let _ = self.tx.take().expect("known to exist").send(());
+                    return true;
+                }
+            }
+            WaitTs::GreaterThan(ts) => {
+                if frontier > ts {
+                    let _ = self.tx.take().expect("known to exist").send(());
+                    return true;
+                }
+            }
+        };
+
+        if let Some(tx) = self.tx.as_ref() {
+            if tx.is_closed() {
+                // Receiver dropped, so also complete.
+                self.tx.take();
+                return true;
+            }
+        }
+
+        false
+    }
+}
+
+impl<T> TxnsReadTask<T>
 where
     T: Timestamp + Lattice + TotalOrder + StepForward + Codec64,
-    C: TxnsCodec,
 {
     async fn run(&mut self) {
         while let Some(cmd) = self.rx.recv().await {
             match cmd {
+                TxnsReadCmd::Updates { entries, frontier } => {
+                    tracing::trace!(
+                        "updates from subscribe task at ({:?}): {:?}",
+                        frontier,
+                        entries
+                    );
+
+                    self.cache.push_entries(entries, frontier.clone());
+
+                    // Respond to waits and retain those that still have to
+                    // wait.
+                    self.pending_waits.retain_mut(|wait| {
+                        let completed = wait.maybe_complete(&frontier);
+
+                        !completed
+                    });
+
+                    self.frontier = frontier;
```

Comment:
Oh missed this on the first pass. Isn't this `self.frontier` redundant with the one in the cache? I'd lean toward exposing that one (if necessary) and using it instead of the second copy

Reply:
yep! it's already `pub(crate)` :sweat_smile: